runtime/autoload/provider/node.vim

```diff
@@ -41,17 +57,11 @@ function! provider#node#Detect() abort
   if exists('g:node_host_prog')
     return g:node_host_prog
   endif
-  let global_modules = get(split(system('npm root -g'), "\n"), 0, '')
-  if v:shell_error || !isdirectory(global_modules)
-    return ''
-  endif
   if !s:is_minimum_version(v:null, 6, 0)
     return ''
   endif
-  let entry_point = glob(global_modules . '/neovim/bin/cli.js')
-  if !filereadable(entry_point)
-    return ''
-  endif
+  let entry_point = s:get_node_client_entry('npm root -g')
+  let entry_point = !empty(entry_point) ? entry_point : s:get_node_client_entry('yarn global dir')
```

Comment:
Pass the name of the package manager instead to determine the command for the directory for global node modules. It's easier to check if `npm` or `yarn` is in `PATH` by checking if they're `executable()`.

Reply:
We do a check for executables here already https://github.com/neovim/neovim/pull/8528/files#diff-2d7f214e0c52c9cb5f1601bdfe0e46b6R505

but also if I pass only the name then I'll need to reconstruct the command inside the function too depends on that 
```vim
let package_manager = executable(a:package_manager) ? a:package_manager : ''
let cmd =  package_manager ==# 'npm' ? package_manager . ' root -g' : package_manager . ' global dir'
```

Reply:
That's the point. It's internal to the function so that it's easier to use and modify.

Reply:
> That's the point. It's internal to the function so that it's easier to use and modify.

I understand, just thought passing the full command will make it more flexible. So if a new package manager X is released you pass `X find-root` command instead of adding more logic. But I'm fine doing this too.

Reply:
It doesn't hurt to have "redundant" `executable()` checks for autoloaded functions because the user can test each function if they work without running `:checkhealth`.

Reply:
@janlazo what do you think?

```diff
diff --git a/runtime/autoload/provider/node.vim b/runtime/autoload/provider/node.vim
index 2e22e7b46..e86c6a4c9 100644
--- a/runtime/autoload/provider/node.vim
+++ b/runtime/autoload/provider/node.vim
@@ -22,12 +22,16 @@ function! s:is_minimum_version(version, min_major, min_minor) abort
     \         && str2nr(v_list[1]) >= str2nr(a:min_minor)))
 endfunction
 
-function! s:get_node_client_entry(cmd) abort
+function! s:get_node_client_entry(package_manager) abort
+  if !executable(a:package_manager)
+    return ''
+  endif
   let cli_bin = '/neovim/bin/cli.js'
   " `yarn global dir` returns the root directory of the global modules
   " so we need to append the `/node_modules` part
-  let yarn_extra_path = a:cmd ==# 'yarn global dir' ? '/node_modules' : ''
-  let global_modules_dir = get(split(system(a:cmd), "\n"), 0, '') . yarn_extra_path
+  let yarn_extra_path = a:package_manager ==# 'yarn' ? '/node_modules' : ''
+  let cmd =  a:package_manager ==# 'npm' ? a:package_manager . ' root -g' : a:package_manager . ' global dir'
+  let global_modules_dir = get(split(system(cmd), "\n"), 0, '') . yarn_extra_path
   if v:shell_error || !isdirectory(global_modules_dir)
     return ''
   endif
@@ -60,8 +64,8 @@ function! provider#node#Detect() abort
   if !s:is_minimum_version(v:null, 6, 0)
     return ''
   endif
-  let entry_point = s:get_node_client_entry('npm root -g')
-  let entry_point = !empty(entry_point) ? entry_point : s:get_node_client_entry('yarn global dir')
+  let entry_point = s:get_node_client_entry('npm')
+  let entry_point = !empty(entry_point) ? entry_point : s:get_node_client_entry('yarn')
   return entry_point
 endfunction
```

Reply:
You need to check if `yarn global dir` returns a valid path on top of its `node_modules/` subfolder otherwise it might use `/node_modules/`, where `/` is root, as its entry point.

Can you make the code more explicit? Only npm and yarn are supported.

```vim
let cmd = a:package_manager ==# 'yarn' ? 'yarn global dir' : 'npm root -g'
```

Reply:
> You need to check if `yarn global dir` returns a valid path on top of its `node_modules/` subfolder otherwise it might use `/node_modules/`, where `/` is root, as its entry point if `v:shell_error` returns 0.

The check is already here, no? `if v:shell_error || !isdirectory(global_modules_dir)`
https://github.com/neovim/neovim/pull/8528/files/096a99f3dee634c11843baeb0fb3cc724d81fda4#diff-f9dd34ad99ab738d7adc7261f3fbb879R31

> Can you make the code more explicit? Only npm and yarn are supported.
>
> ```vim
> let cmd = a:package_manager ==# 'yarn' ? 'yarn global dir' : 'npm root -g'
> ```

Ok will do that.

Reply:
@janlazo good now?

```vim
function! s:get_node_client_entry(package_manager) abort
  if !executable(a:package_manager)
    return ''
  endif
  let cli_bin = '/neovim/bin/cli.js'
  " `yarn global dir` returns the root directory of the global modules
  " so we need to append the `/node_modules` part
  let yarn_extra_path = a:package_manager ==# 'yarn' ? '/node_modules' : ''
  let cmd =  a:package_manager ==# 'yarn' ? 'yarn global dir' : 'npm root -g'
  let global_modules_dir = get(split(system(cmd), "\n"), 0, '') . yarn_extra_path
  if v:shell_error || !isdirectory(global_modules_dir)
    return ''
  endif
  let entry_path = glob(global_modules_dir . cli_bin)
  if !filereadable(entry_path)
    return ''
  endif
  return entry_path
endfunction


function! provider#node#Detect() abort
  if exists('g:node_host_prog')
    return g:node_host_prog
  endif
  if !s:is_minimum_version(v:null, 6, 0)
    return ''
  endif
  let entry_point = s:get_node_client_entry('npm')
  let entry_point = !empty(entry_point) ? entry_point : s:get_node_client_entry('yarn')
  return entry_point
endfunction
```

Reply:
Why did you break up `get(split(system(cmd), "\n"), 0, '')`? You need that to remove traililng `<CR>` before passing the output to `isdirectory`.

Reply:
That was a mistake, this is better now?

```vim
function! s:get_node_client_entry(package_manager) abort
  if !executable(a:package_manager)
    return ''
  endif
  let cli_bin = '/neovim/bin/cli.js'
  let cmd =  a:package_manager ==# 'yarn' ? 'yarn global dir' : 'npm root -g'
  let modules_dir = get(split(system(cmd), "\n"), 0, '')
  if v:shell_error || !isdirectory(modules_dir)
    return ''
  endif
  " `yarn global dir` returns the root directory of the global modules
  " so we need to append the `/node_modules` part
  let global_modules_dir = a:package_manager ==# 'yarn' ? modules_dir . '/node_modules' : modules_dir
  if !isdirectory(global_modules_dir)
    return ''
  endif
  let entry_path = glob(global_modules_dir . cli_bin)
  if !filereadable(entry_path)
    return ''
  endif
  return entry_path
endfunction



function! provider#node#Detect() abort
  if exists('g:node_host_prog')
    return g:node_host_prog
  endif
  if !s:is_minimum_version(v:null, 6, 0)
    return ''
  endif
  let entry_point = s:get_node_client_entry('npm')
  let entry_point = !empty(entry_point) ? entry_point : s:get_node_client_entry('yarn')
  return entry_point
endfunction
```

Reply:
@janlazo I pushed the latest changes have a look & let me know what you think.